### DISCUSSION TITLE
Support for GLTF File Format

### DISCUSSION THREAD

**Discussion by rooomFelix** [BrunoLevy/geogram#158] [Ideas]
####  Summary
I would like to request the addition of support for the GLTF (GL Transmission Format) file format in the Geogram. This enhancement would greatly benefit users by expanding the versatility and usability of the library in modern 3D graphics workflows.

#### Detailed Description
The GLTF file format, developed by the Khronos Group, has rapidly become the standard for 3D model transmission and interchange. Its efficient, compact representation and compatibility with modern web technologies make it an ideal choice for a wide range of applications, including virtual reality (VR), augmented reality (AR), and real-time rendering.

Key benefits of supporting GLTF in the Geogram include:
1. **Enhanced Compatibility**: Many 3D content creation tools and engines (such as Blender, Unity, and Unreal Engine) natively support GLTF. By integrating GLTF support, the library would seamlessly fit into existing pipelines, reducing the need for additional conversion steps.
2. **Performance Optimization**: GLTF is designed for efficient transmission and loading, enabling faster model delivery and improved performance in web and mobile applications. The library can leverage this efficiency to optimize 3D models without sacrificing quality.
3. **Wide Adoption**: As an open standard, GLTF is widely adopted in the industry. Supporting it would attract a broader user base to the Geogram, fostering a more extensive and active community.
4. **Rich Feature Set**: GLTF supports PBR (Physically Based Rendering) materials, animations, skins, morph targets, and more. Incorporating these features into the optimization process would enhance the visual fidelity and functionality of the resulting models.

Best regards,
Felix

**Comment by BrunoLevy** [comment]
Hello,
I have no short-term plan for adding GLTF, but I agree this could be interesting.

In the meanwhile, what you can do is developing a add-on: it is easy to register new file formats for geogram,
you just need write a class that inherits [MeshIOHandler](https://brunolevy.github.io/geogram/classGEO_1_1MeshIOHandler.html) and declare it to the system using [geo_register_MeshIOHandler_creator](https://brunolevy.github.io/geogram/mesh__io_8h.html#a2af344a055948ce3a4268bf177f966ef).

See [mesh/mesh_IO.cpp](https://github.com/BrunoLevy/geogram/blob/2fe5307cab930dfb161d2bee7711ad4cb65368c6/src/lib/geogram/mesh/mesh_io.cpp#L5023) for examples.

**Reply by rooomFelix** [reply]
Hello Bruno,

Thank you for your prompt feedback. Yes, I have already reviewed the classes. Unfortunately, I am "only" a JavaScript frontend developer, which explains my somewhat audacious request :) The possibility of using Geogram as a WASM library is very exciting, and this is why I am particularly interested in its application within a WebGL context (and related Frameworks like three.js, babylon.js) and its direct use in the browser.

I look forward to seeing this feature on the roadmap in the future. Your library is very intriguing, and I am excited about its development.

**Comment by sebmestrallet** [comment]
Hello Felix,

I needed to export glTF assets from my Geogram-based project, and I used the [tinygltf](https://github.com/syoyo/tinygltf) C++ lib from [@syoyo](https://github.com/syoyo), which API is very close to the glTF spec.

I you want to develop a glTF writer in C++ by assembling the fields from a `GEO::Mesh` object, I can send you a code snippet.

However, from a web perspective, Javascript seems sufficient to handle & render glTF assets, see [modelviewer.dev](https://modelviewer.dev/) ([google/model-viewer](https://github.com/google/model-viewer)) or [https://3dviewer.net](https://3dviewer.net/) ([kovacsv/Online3DViewer](https://github.com/kovacsv/Online3DViewer))

**Reply by ftc26** [reply]
Hello Sébastien,
Thank you for the offer. I will gladly accept it.
I want to use some of the features of geogram like decimate and remeshing. I am currently using https://gltf-transform.dev/ to modify gltf data and geogram would be a great addition.
Thank you!

**Reply by sebmestrallet** [reply]
From what I understand, your pipeline requires a significant modification of Geogram.

You have an input glTF (with a texture?) that you are editing with glTF-Transform (keeping the attributes I guess), then you want to process it further with Geogram, and export the final glTF. If you only need the mesh, not the texture, the easier would be to extract the mesh from the glTF as .obj, then process the .obj with Geogram. However, if you need the whole pipeline to preserve texture info, Geogram will need a `MeshIOHandler` that map glTF-specific data ↔ Geogram attributes, and you will need to ensure all Geogram algorithms you are using are preserving attributes.

Here a basic glTF writer with the following limits:
- only for triangle meshes
- no texture
- 32 bits indices
- double precision coordinates as input, but single precision coordinates written
- binary output (.glb)

```cpp
void write_triangle_mesh_as_glTF(std::string filename, Mesh& M, bool with_wireframe) {
	// assert Gargantua mode is off
	geo_assert(sizeof(index_t)==4); // 4 bytes = 32 bits
    
    ////////////////////////
    // Get wireframe edges
    ////////////////////////

    std::set<std::set<index_t>> wireframe_edges_as_set;
    if(with_wireframe) {
        GEO::MeshHalfedges::Halfedge H;
        index_t H_v0 = index_t(-1);
        index_t H_v1 = index_t(-1);
        FOR(f,M.facets.nb()) { // for each facet
            H.facet = f;
            FOR(lv,3) { // for each local vertex of f
                H.corner = M.facets.corner(f,lv);
                H_v0 = halfedge_vertex_index_from(M,H); // see PR #116 
                H_v1 = halfedge_vertex_index_to(M,H); // see PR #116 
                wireframe_edges_as_set.insert({H_v0,H_v1});
            }
        }
    }
    // we need to have an index associated to each wireframe edge
    // -> convert to vector
    std::vector<std::set<index_t>> wireframe_edges_as_vector(wireframe_edges_as_set.begin(),wireframe_edges_as_set.end());

    /////////////////////////
    // Compute bounding box
    /////////////////////////

    // /!\ warning: vertices coordinates must be double precision
    std::vector<double> bounding_box_min(3,0.0), bounding_box_max(3,0.0);
    GEO::get_bbox(M, bounding_box_min.data(), bounding_box_max.data());

    //////////////////////////////////////////
    // Convert vertices coordinates to float
    //////////////////////////////////////////

    M.vertices.set_single_precision();

    ////////////////////////
    // Create a glTF model
    ////////////////////////

    tinygltf::Model m;

    //////////////////////////////////////////////
    // Create 1 material (+1 with the wireframe)
    //////////////////////////////////////////////

    m.materials.resize(1);
    const size_t MATERIAL_0_GREY = 0;
    const size_t MATERIAL_1_BLACK = 1;
    tinygltf::Material& material_0_grey = m.materials[MATERIAL_0_GREY];
    
    material_0_grey.pbrMetallicRoughness.baseColorFactor = {0.8f, 0.8f, 0.8f, 1.0f};
    material_0_grey.doubleSided = true;

    if(with_wireframe) {
        m.materials.resize(2);
        tinygltf::Material& material_1_black = m.materials[MATERIAL_1_BLACK];
        material_1_black.pbrMetallicRoughness.baseColorFactor = {0.0f, 0.0f, 0.0f, 1.0f};
        material_1_black.doubleSided = true;
    }

    ////////////////////
    // Create a buffer
    ////////////////////

    m.buffers.resize(1);
    const size_t BUFFER_0 = 0;
    tinygltf::Buffer& buffer_0 = m.buffers[BUFFER_0];

    // resize buffer to 4 bytes * 3 floating points * #vertices
    //                + 4 bytes * 3 indices * #triangles
    //                + 4 bytes * 2 indices * #edges
    size_t buffer_coordinates_start         = 0;                                                // byte index
    size_t buffer_coordinates_length        = 4*3*M.vertices.nb();                              // number of bytes
    size_t buffer_triangles_start           = buffer_coordinates_length;                        // byte index
    size_t buffer_triangles_length          = 4*3*M.facets.nb();                                // number of bytes
    size_t buffer_wireframe_edges_start     = buffer_triangles_start+buffer_triangles_length;   // byte index
    size_t buffer_wireframe_edges_length    = 4*2*wireframe_edges_as_vector.size();             // number of bytes
    if(with_wireframe) {
        buffer_0.data.resize(buffer_coordinates_length+buffer_triangles_length+buffer_wireframe_edges_length);
    }
    else {
        buffer_0.data.resize(buffer_coordinates_length+buffer_triangles_length);
    }

    // write vertices (= 3D coordinates)
    FOR(v,M.vertices.nb()) { // for each vertex index
        memcpy(buffer_0.data.data() + v*12, static_cast<void*>(M.vertices.single_precision_point_ptr(v)), 12); // write x,y,z from float* getter
    }
    // write triangles (= vertex indices)
    FOR(f,M.facets.nb()) { // for each facet (triangle) index
        memcpy(buffer_0.data.data() + buffer_triangles_start + f*12 + 0, facet_vertex_index_ptr(M,f,0),4);
        memcpy(buffer_0.data.data() + buffer_triangles_start + f*12 + 4, facet_vertex_index_ptr(M,f,1),4);
        memcpy(buffer_0.data.data() + buffer_triangles_start + f*12 + 8, facet_vertex_index_ptr(M,f,2),4);
    }
    if(with_wireframe) {
        // write wireframe edges (= vertex indices)
        FOR(e,wireframe_edges_as_vector.size()) {
            memcpy(buffer_0.data.data() + buffer_wireframe_edges_start + 8*e + 0, &(*wireframe_edges_as_vector[e].begin()),  4);
            memcpy(buffer_0.data.data() + buffer_wireframe_edges_start + 8*e + 4, &(*wireframe_edges_as_vector[e].rbegin()), 4);
        }
    }    

    //////////////////////////////////////////////////
    // Create 2 buffer views (+1 with the wireframe)
    //////////////////////////////////////////////////

    m.bufferViews.resize(2);
    const size_t BUFFERVIEW_0_VERTICES_COORDINATES = 0;
    const size_t BUFFERVIEW_1_TRIANGLES_VERTICES = 1;
    const size_t BUFFERVIEW_2_WIREFRAME_EDGES_VERTICES = 2;
    tinygltf::BufferView& bufferview_0_vertices_coordinates = m.bufferViews[BUFFERVIEW_0_VERTICES_COORDINATES];
    tinygltf::BufferView& bufferview_1_triangles_vertices = m.bufferViews[BUFFERVIEW_1_TRIANGLES_VERTICES];
    

    // 1st chunk of the buffer 0 : points = 3D coordinates (type ARRAY_BUFFER)
    bufferview_0_vertices_coordinates.buffer = BUFFER_0;
    bufferview_0_vertices_coordinates.byteOffset = buffer_coordinates_start;
    bufferview_0_vertices_coordinates.byteLength = buffer_coordinates_length;
    bufferview_0_vertices_coordinates.target = TINYGLTF_TARGET_ARRAY_BUFFER;

    // 2nd chunk of the buffer 0 : triangles = vertex indices (type ELEMENT_ARRAY_BUFFER)
    bufferview_1_triangles_vertices.buffer = BUFFER_0;
    bufferview_1_triangles_vertices.byteOffset = buffer_triangles_start;
    bufferview_1_triangles_vertices.byteLength = buffer_triangles_length;
    bufferview_1_triangles_vertices.target = TINYGLTF_TARGET_ELEMENT_ARRAY_BUFFER;

    if(with_wireframe) {
        m.bufferViews.resize(3);
        tinygltf::BufferView& bufferview_2_wireframe_edges_vertices = m.bufferViews[BUFFERVIEW_2_WIREFRAME_EDGES_VERTICES];
        // 3rd chunk of the buffer 0 : edges = vertex indices (type ELEMENT_ARRAY_BUFFER)
        bufferview_2_wireframe_edges_vertices.buffer = BUFFER_0;
        bufferview_2_wireframe_edges_vertices.byteOffset = buffer_wireframe_edges_start;
        bufferview_2_wireframe_edges_vertices.byteLength = buffer_wireframe_edges_length;
        bufferview_2_wireframe_edges_vertices.target = TINYGLTF_TARGET_ELEMENT_ARRAY_BUFFER;
    }

    ///////////////////////////////////////////////
    // Create 2 accessors (+1 with the wireframe)
    ///////////////////////////////////////////////

    m.accessors.resize(2);
    const size_t ACCESSOR_0_VERTICES_COORDINATES = 0;
    const size_t ACCESSOR_1_TRIANGLES_VERTICES = 1;
    const size_t ACCESSOR_2_WIREFRAME_EDGES_VERTICES = 2;
    tinygltf::Accessor& accessor_0_vertices_coordinates = m.accessors[ACCESSOR_0_VERTICES_COORDINATES];
    tinygltf::Accessor& accessor_1_triangles_vertices = m.accessors[ACCESSOR_1_TRIANGLES_VERTICES];

    // layout description of buffer view 0
    accessor_0_vertices_coordinates.bufferView = BUFFERVIEW_0_VERTICES_COORDINATES;
    accessor_0_vertices_coordinates.byteOffset = 0;
    accessor_0_vertices_coordinates.componentType = TINYGLTF_COMPONENT_TYPE_FLOAT;
    accessor_0_vertices_coordinates.count = M.vertices.nb(); // how many points
    accessor_0_vertices_coordinates.type = TINYGLTF_TYPE_VEC3;
    accessor_0_vertices_coordinates.maxValues = bounding_box_max; // = [max_x, max_y, max_z]
    accessor_0_vertices_coordinates.minValues = bounding_box_min; // = [min_x, min_y, min_z]

    // layout description of buffer view 1
    accessor_1_triangles_vertices.bufferView = BUFFERVIEW_1_TRIANGLES_VERTICES;
    accessor_1_triangles_vertices.byteOffset = 0;
    accessor_1_triangles_vertices.componentType = TINYGLTF_COMPONENT_TYPE_UNSIGNED_INT; // 32 bits per index
    accessor_1_triangles_vertices.count = M.facets.nb()*3; // how many indices
    accessor_1_triangles_vertices.type = TINYGLTF_TYPE_SCALAR;
    // range of indices value : [ 0 : #vertices-1 ]
    accessor_1_triangles_vertices.maxValues.push_back(M.vertices.nb()-1); // max value of all indices in the buffer view
    accessor_1_triangles_vertices.minValues.push_back(0);

    if(with_wireframe) {
        m.accessors.resize(3);
        tinygltf::Accessor& accessor_2_wireframe_edges_vertices = m.accessors[ACCESSOR_2_WIREFRAME_EDGES_VERTICES];
        // layout description of buffer view 2
        accessor_2_wireframe_edges_vertices.bufferView = BUFFERVIEW_2_WIREFRAME_EDGES_VERTICES;
        accessor_2_wireframe_edges_vertices.byteOffset = 0;
        accessor_2_wireframe_edges_vertices.componentType = TINYGLTF_COMPONENT_TYPE_UNSIGNED_INT; // 32 bits per index
        accessor_2_wireframe_edges_vertices.count = wireframe_edges_as_vector.size()*2; // how many indices
        accessor_2_wireframe_edges_vertices.type = TINYGLTF_TYPE_SCALAR;
        // range of indices value : [ 0 : #vertices-1 ]
        accessor_2_wireframe_edges_vertices.maxValues.push_back(M.vertices.nb()-1); // max value of all indices in the buffer view
        accessor_2_wireframe_edges_vertices.minValues.push_back(0);
    }

    ///////////////////////////////////////////////////////////
    // Create a mesh with 1 primitive (+1 with the wireframe)
    ///////////////////////////////////////////////////////////

    m.meshes.resize(1);
    const size_t MESH_0 = 0;
    tinygltf::Mesh& mesh_0 = m.meshes[MESH_0];

    mesh_0.primitives.resize(1);
    const size_t PRIMITIVE_0_TRIANGLES = 0;
    const size_t PRIMITIVE_1_WIREFRAME_EDGES = 1;
    tinygltf::Primitive& primitive_0_triangles = mesh_0.primitives[PRIMITIVE_0_TRIANGLES];

    primitive_0_triangles.indices = ACCESSOR_1_TRIANGLES_VERTICES;
    primitive_0_triangles.attributes["POSITION"] = ACCESSOR_0_VERTICES_COORDINATES;
    primitive_0_triangles.material = MATERIAL_0_GREY;
    primitive_0_triangles.mode = TINYGLTF_MODE_TRIANGLES;

    if(with_wireframe) {
        mesh_0.primitives.resize(2);
        tinygltf::Primitive& primitive_1_wireframe_edges = mesh_0.primitives[PRIMITIVE_1_WIREFRAME_EDGES];
        primitive_1_wireframe_edges.indices = ACCESSOR_2_WIREFRAME_EDGES_VERTICES;
        primitive_1_wireframe_edges.attributes["POSITION"] = ACCESSOR_0_VERTICES_COORDINATES;
        primitive_1_wireframe_edges.material = MATERIAL_1_BLACK;
        primitive_1_wireframe_edges.mode = TINYGLTF_MODE_LINE;
    }

    //////////////////
    // Create a node
    //////////////////
    
    m.nodes.resize(1);
    const size_t NODE_0 = 0;

    m.nodes[NODE_0].mesh = MESH_0; // link NODE_0 to MESH_0

    ///////////////////
    // Create a scene
    ///////////////////

    m.scenes.resize(1);
    const size_t SCENE_0 = 0;

    m.scenes[SCENE_0].nodes.push_back(NODE_0); // link SCENE_0 to NODE_0

    ///////////////////////
    // Describe the asset
    ///////////////////////

    m.asset.version = "2.0"; // required
    m.asset.generator = "tinygltf";
    
    ///////////////////////
    // Save model to file
    ///////////////////////

    tinygltf::TinyGLTF gltf;
    fmt::println(Logger::out("glTF"),"Writing {}...",filename); Logger::out("glTF").flush(); // or without the {fmt} lib if you like << streams <<
    gltf.WriteGltfSceneToFile(&m, filename,
                            true, // embedImages
                            true, // embedBuffers
                            true, // pretty print
                            true); // write binary
}
```

Et voilà, you can have a .glb bunny : 

![Capture d’écran du 2024-09-05 13-54-45](https://github.com/user-attachments/assets/c684d057-f8ba-406e-acfd-29a5326f092a)

I have some code to export a colored triangle mesh, but it is very tailored for my needs.

**Reply by ftc26** [reply]
Yes, you're right. With the gltf format there are some attributes that should be retained. bone weights, morph weights, uvs, ... Maybe the mapping of this information becomes too complex. I usually use glb files with integrated textures.

**Reply by sebmestrallet** [reply]
As far as I know, Geogram meshes only store vertices, facets and cells (+ independent edges), and arbitrary attributes for any of these element as well as for facet corners, cell corners and cell facets.

UVs can be stored as vertex attributes, but bones & morph targets need new data structures I think + to be taken into account by Geogram's algorithms

**Reply by ftc26** [reply]
ok then it may not be possible if it needs other data structures.
I thought that for each morph channel in a gltf a separate vertices attribute channel is created where the weight are stored as floats. the same for bones.

**Reply by sebmestrallet** [reply]
Each morph channel stores a separate vertices coordinates list, but weights are not attached to vertices; they are in an independent list, of length nb_channels, see [tutorial 018](https://github.com/KhronosGroup/glTF-Tutorials/blob/main/gltfTutorial/gltfTutorial_018_MorphTargets.md). For an animation (variable weights), each time step stores a list of weights value.

**Reply by BrunoLevy** [reply]
It would be possible to store the bones in the edges of the mesh and the weights in attributes, but the best way of storing everything is probably to have separate meshes for the bones. This could be implemented as a general function that returns several meshes + the images, and a more minimalistic MeshIO binding that just reads the mesh.

**Reply by ftc26** [reply]
So i wonder if it is possible to simply transfer features like morph and animation channels. Even if the topology of the mesh changes due to decimate, these attributes have to be adapted. Looks like a more complex task.

**Reply by BrunoLevy** [reply]
I think that blending weights can be transferred, there might be some subtlety near the joints though ...
Functions to transfer attributes between different meshes are in [`geogram/mesh/mesh_baking.h`](https://brunolevy.github.io/geogram/mesh__baking_8h.html)
In general, transferring structured information between two different meshes is challenging.

**Reply by tksuoran** [reply]
> As far as I know, Geogram meshes only store vertices, facets and cells (+ independent edges), and arbitrary attributes for any of these element as well as for facet corners, cell corners and cell facets.
> 
> UVs can be stored as vertex attributes, but bones & morph targets need new data structures I think + to be taken into account by Geogram's algorithms

While bones are indeed external, joint indices and joint weights are vertex attributes.

**Comment by BrunoLevy** [comment]
Hello Sébastien,
Interesting ! If you post the code snippet I'll take a look and see how we could integrate it.

**Comment by tksuoran** [comment]
I have glTF loading code (using fastgltf for parsing), and converts it to simple(ish) triangle soup data structure. This includes support for joint indices and joint weights The triangle soup data structure should be easy to convert to `GEO::Mesh`. I have code which does this, but currently it goes through yet another intermediate data structure: My code uses inferior `GEO::Mesh` -like data structure, which is capable of containing Mesh facets, vertices, corners, and attributes for these. I am in the process of replacing this inferior data structure with `GEO::Mesh`. This change will take some time, as I do it on my spare time. When I am done, I will have capability to load glTF into `GEO::Mesh`.

IMHO, the best way to deal features (such as skeletons) not supported by Geogram, is to keep them as-is from the input glTF. Once processing is done with Geogram, only the mesh data (that ends up in index and vertex buffers) needs to be replaced; rest of the glTF input could be passed through without modifications to output.
